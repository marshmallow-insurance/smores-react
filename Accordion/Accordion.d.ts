import { FC, ReactNode } from 'react';
import { MarginProps } from '../utils/space';
export type AccordionProps = {
    title: string;
    subTitle?: string;
    filledBackground?: boolean;
    borderTop?: boolean;
    borderColor?: 'oatmeal' | 'custard' | 'cream' | 'coconut';
    fullBorder?: boolean;
    backgroundColor?: 'oatmeal' | 'custard' | 'cream' | 'coconut';
    onToggle?: (isOpen: boolean) => void;
    children: ReactNode;
    defaultIsOpen?: boolean;
} & MarginProps;
export declare const Accordion: FC<AccordionProps>;
