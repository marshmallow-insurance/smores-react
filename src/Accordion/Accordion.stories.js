import React from 'react'
import { Accordion } from './Accordion'

export default {
  title: 'Accordion',
  component: Accordion,
}

const Template = (args) => (
  <Accordion {...args} title="How it works">
    Lots of brilliant information about this beautiful component
  </Accordion>
)

export const Default = Template.bind({})
