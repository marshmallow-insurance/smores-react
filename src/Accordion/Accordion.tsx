import React, { FC, useState } from 'react'
import styled, { css } from 'styled-components'

import { Box } from '../Box'
import { Icon } from '../Icon'
import { theme } from '../theme'
import { Text } from '../Text'
import { MarginProps } from '../utils/space'

export type AccordionProps = {
  title: string
  subTitle?: string
  borderTop?: boolean
  fullBorder?: boolean
  onToggle?: VoidFunction
} & MarginProps

export const Accordion: FC<AccordionProps> = ({
  title,
  children,
  onToggle,
  borderTop = false,
  subTitle,
  fullBorder = false,
  ...marginProps
}) => {
  const [isOpen, setIsOpen] = useState(false)
  const px = fullBorder ? '16px' : '0'

  const handleToggle = () => {
    onToggle?.()
    setIsOpen(!isOpen)
  }

  return (
    <Wrapper borderTop={borderTop} fullBorder={fullBorder} {...marginProps}>
      <TopContainer
        flex
        alignItems="center"
        justifyContent="space-between"
        onClick={handleToggle}
        py={fullBorder ? { custom: 20 } : '16px'}
        px={px}
      >
        <TitleContainer>
          <Text tag="h2" typo="headline-regular">
            {title}
          </Text>
          {subTitle && (
            <Text tag="label" color="subtext" typo="label">
              {subTitle}
            </Text>
          )}
        </TitleContainer>

        <CaretIcon
          render="caret"
          size={24}
          color="secondary"
          isOpen={isOpen}
          borderTop={borderTop}
        />
      </TopContainer>
      {isOpen && (
        <Box pt="12px" pb="16px" px={px}>
          {children}
        </Box>
      )}
    </Wrapper>
  )
}

interface IAccordion {
  isOpen: boolean
  borderTop?: boolean
  fullBorder?: boolean
}

const Wrapper = styled(Box)<Omit<IAccordion, 'isOpen'>>(
  ({ borderTop, fullBorder }) => css`
    border-bottom: 1px solid ${theme.colors.outline};
    ${borderTop && `border-top: 1px solid ${theme.colors.outline};`}

    ${fullBorder &&
    css`
      border: 1px solid ${theme.colors.outline};
      border-radius: 8px;
      margin-bottom: 14px;
    `}
  `,
)

const TitleContainer = styled.div`
  display: flex;
  flex-direction: column;
`
const TopContainer = styled(Box)`
  cursor: pointer;
`

const CaretIcon = styled(Icon)<IAccordion>(
  ({ isOpen }) => css`
    transform: ${isOpen ? 'rotate(180deg)' : 'rotate(0deg)'};
    transition: transform 0.6s ease;
  `,
)
