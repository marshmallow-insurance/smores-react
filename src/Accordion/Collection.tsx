import React, { FC, ReactNode } from 'react'
import styled from 'styled-components'

import { Text } from '../Text'
import { Box } from '../Box'
import { Accordion, AccordionProps } from './Accordion'

const accordionList: Array<AccordionProps> = [
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
  },
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
    subTitle: 'subtitle',
  },
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
    filledBackground: true,
  },
  {
    children: 'Lots of brilliant information about this beautiful component',
    title: 'How it works',
    subTitle: 'subtitle',
    filledBackground: true,
  },
]

export const CollectionPage: FC = () => {
  return (
    <Wrapper>
      <Section label="Default">
        {accordionList.map((props, index) => (
          <Accordion key={index} {...props} />
        ))}
      </Section>
      <Section label="Border Top">
        {accordionList.map((props, index) => (
          <Accordion key={index} {...props} borderTop={true} />
        ))}
      </Section>
      <Section label="Full Border">
        {accordionList.map((props, index) => (
          <Accordion key={index} {...props} fullBorder={true} />
        ))}
      </Section>
    </Wrapper>
  )
}

const Wrapper = styled(Box)`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 40px;
  width: 100%;
`

const Section: FC<{ label: string; children?: ReactNode }> = ({
  label,
  children,
}) => {
  return (
    <Box width="100%" mb="16px">
      <Box mb="16px">
        <Text tag="span" typo="header-small">
          {label}
        </Text>
      </Box>
      <Column>{children}</Column>
    </Box>
  )
}

const Column = styled(Box)`
  display: flex;
  flex-direction: column;
  gap: 16px;
`
