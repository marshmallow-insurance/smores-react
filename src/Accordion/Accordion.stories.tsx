import React from 'react'
import { Accordion, AccordionProps } from './Accordion'
import { CollectionPage } from './Collection'

export default {
  title: 'Accordion',
  component: Accordion,
}

const Template = (props: AccordionProps) => (
  <Accordion {...props} title="How it works">
    Lots of brilliant information about this beautiful component
  </Accordion>
)

export const Default = Template.bind({})

export const Collection = CollectionPage.bind({})
