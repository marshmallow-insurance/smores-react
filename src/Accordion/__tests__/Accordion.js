import React from 'react'
import { render } from '@testing-library/react'
import 'jest-styled-components'

import { Accordion } from '../Accordion'

test('renders', () => {
  const { container } = render(
    <Accordion title="How it works">
      Lots of brilliant information about this beautiful component
    </Accordion>,
  )
  expect(container.firstChild).toMatchSnapshot()
})
