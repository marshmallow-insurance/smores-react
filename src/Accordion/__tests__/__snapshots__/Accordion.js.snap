// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c4 {
  font-size: 16px;
  font-weight: 500;
  line-height: 20px;
  text-align: left;
  cursor: inherit;
  color: #292924;
}

.c0 {
  border-bottom: 1px solid #DAD2C4;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  cursor: pointer;
}

.c6 {
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  -webkit-transition: -webkit-transform 0.6s ease;
  -webkit-transition: transform 0.6s ease;
  transition: transform 0.6s ease;
}

@media (min-width:0px) {
  .c1 {
    padding-top: 16px;
    padding-bottom: 16px;
  }
}

@media (min-width:0px) {
  .c1 {
    padding-left: 0;
  }
}

@media (min-width:0px) {
  .c1 {
    padding-right: 16px;
  }
}

@media (min-width:0px) {
  .c1 {
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
  }
}

@media (min-width:0px) {
  .c1 {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
  }
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <div
      class="c3"
    >
      <h2
        class="c4"
        color="liquorice"
        cursor="inherit"
        title=""
      >
        How it works
      </h2>
    </div>
    <span
      class="c5 c6"
      color="marzipan"
      rotate="0"
      size="24"
    >
      <svg
        fill="none"
        height="100%"
        viewBox="0 0 24 24"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19.66 8.248a1 1 0 00-1.412.095L12 15.482l-6.248-7.14a1 1 0 10-1.504 1.317l7 8a.995.995 0 001.504 0l7-8a1 1 0 00-.093-1.411z"
          fill="#A1957C"
        />
      </svg>
    </span>
  </div>
</div>
`;
